third_party/terraform/utils/cloudrun_polling.go

```diff
@@ -31,91 +30,39 @@ type KnativeStatus struct {
 	}
 }
 
-// ConditionByType is a helper method for extracting a given condition
-func (s KnativeStatus) ConditionByType(typ string) *Condition {
-	for _, condition := range s.Status.Conditions {
-		if condition.Type == typ {
-			c := condition
-			return &c
-		}
-	}
-	return nil
-}
+type KNativeStatusPoll struct{}
 
-// LatestMessage will return a human consumable status of the resource. This can
-// be used to determine the human actionable error the GET doesn't return an explicit
-// error but the resource is in an error state.
-func (s KnativeStatus) LatestMessage() string {
-	c := s.ConditionByType(readyStatus)
-	if c != nil {
-		return fmt.Sprintf("%s - %s", c.Reason, c.Message)
+func (p *KNativeStatusPoll) ShouldContinuePolling(resp map[string]interface{}, respErr error) (bool, error) {
+	if respErr != nil {
+		return false, respErr
+	}
+	s := KnativeStatus{}
+	err := Convert(resp, &s)
+	if err != nil {
+		return false, errwrap.Wrapf("Unable to convert response into a KnativeStatus: {{err}}", err)
 	}
 
-	return ""
-}
-
-// State will return a string representing the status of the Ready condition.
-// No other conditions are currently returned as part of the state.
-func (s KnativeStatus) State(res interface{}) string {
 	for _, condition := range s.Status.Conditions {
-		if condition.Type == "Ready" {
-			// DomainMapping can enter a 'terminal' state of waiting for external verification
-			// of DNS records.
-			if condition.Reason == "CertificatePending" {
-				return "Ready:CertificatePending"
+		if condition.Type == readyStatusType {
+			log.Printf("[DEBUG] checking KnativeStatus Ready condition %s: %s", condition.Status, condition.Message)
+			switch condition.Status {
+			case "Unknown":
+				// DomainMapping can enter a 'terminal' state where "Ready" status is "Unknown"
+				// but the resource is waiting for external verification of DNS records.
+				if condition.Reason == pendingCertificateReason {
+					return false, nil
+				}
+				return true, fmt.Errorf(`found pending status "Ready:Unknown"`)
```

Comment:
Most of the errors in this method are just status messages. This would require the caller to invert the standard golang checking logic of `if err!= nil, do something` which is a bit of an anti-pattern.